### PR TITLE
Conversion of modules from public to private by default

### PR DESCRIPTION
**Pull request by glemieux**
### Description:
This pull request converts all remaining modules that were public by default to `private` by default.  About 50% of the existing code base was already defined as `private` by default.  The majority of the changes were to modules that are part of `main` with only a few other functions in `parteh` and `fire`.

This work was conducted in response to issue #502.    The intent going forward is to have developers explicitly make `public` the procedures, variables, and data necessary to avoid potential 'leaks' in functionality to users.  That said, in this PR all code should retain the same level of privacy as the baseline code it was tested against; i.e. no effort was given to determining what existing code might benefit from a higher level of privacy that currently exists.

*Notes on formatting*:
- All modules now contain a `private` declaration immediately following `implicit none`.
- Variables and types are declared `public` inline.  For example: `integer, parameter, public :: fates_r8 = selected_real_kind(12)` and `type, public :: fates_history_interface_type`
- Procedures (i.e. functions and subroutines) are defined as public typically at the end of the variable and type definitions at the top of the module file like so: `public :: RKF45`.  Note that many procedures are define as public being `contained` within type, which itself may be public.

### Collaborators:
@rgknox @ekluzek 

### Expectation of Answer Changes:
None

### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the in-code documentation .AND. (the [technical note](https://github.com/NGEET/fates-docs) .OR. the wiki) accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/NGEET/fates/blob/master/CONTRIBUTING.md) document.
- [x] FATES PASS/FAIL regression tests were run
- [ ] If answers were expected to change, evaluation was performed and provided

### Test Results:

CTSM (or) E3SM (specify which) test hash-tag: clm5.0.dev008-105-gbc17668

CTSM (or) E3SM (specify which) baseline hash-tag: clm5.0.dev008-105-gbc17668

FATES baseline hash-tag: sci.1.26.3_api.7.2.0-16-gf1d4bc5

Test Output:
*Test output*: /glade/u/home/glemieux/scratch/clmed-tests/private-public-modules.fates.cheyenne.intel.Cbc17668-Fabb9df6

*Baseline output*: /glade/u/home/glemieux/scratch/clmed-baselines/private-public-modules.fates.cheyenne.intel.Cbc17668-Ff1d4bc5

All tests passing excepting:
```
FAIL ERS_D_Ld5.1x1_brazil.I2000Clm50FatesCruGs.cheyenne_intel.clm-FatesHydro COMPARE_base_rest
FAIL ERS_Ld60.f45_f45_mg37.I2000Clm50FatesCruGs.cheyenne_intel.clm-Fates MEMCOMP Error: Memory usage increase > 10% from baseline
```

### CODE REVIEW

File: fire/SFParamsMod.F90 [hunk 1]
```diff
@@ -7,68 +7,58 @@ module SFParamsMod
...
-   character(len=param_string_length),parameter :: SF_name_SAV = "fates_SAV"
-   character(len=param_string_length),parameter :: SF_name_FBD = "fates_FBD"
-   character(len=param_string_length),parameter :: SF_name_min_moisture = "fates_min_moisture"
-   character(len=param_string_length),parameter :: SF_name_mid_moisture = "fates_mid_moisture"
-   character(len=param_string_length),parameter :: SF_name_low_moisture_Coeff = "fates_low_moisture_Coeff"
-   character(len=param_string_length),parameter :: SF_name_low_moisture_Slope = "fates_low_moisture_Slope"
-   character(len=param_string_length),parameter :: SF_name_mid_moisture_Coeff = "fates_mid_moisture_Coeff"
-   character(len=param_string_length),parameter :: SF_name_mid_moisture_Slope = "fates_mid_moisture_Slope"
-
+   real(r8),public,protected :: SF_val_fdi_a
```
Comment (rgknox) [inline fire/SFParamsMod.F90]: @glemieux , if a variable is defined as protected, does it make it implicitly public?  I poked around on the internet and could not find an answer.

File: main/EDTypesMod.F90 [hunk 2]
```diff
@@ -14,27 +14,35 @@ module EDTypesMod
   use FatesConstantsMod,     only : n_anthro_disturbance_categories
 
   implicit none
+  private               ! By default everything is private
   save
 
-  integer, parameter :: maxPatchesPerSite  = 14   ! maximum number of patches to live on a site
-  integer, parameter :: maxPatchesPerSite_by_disttype(n_anthro_disturbance_categories)  = &
+  ! Make public necessary subroutines and functions
+  public :: val_check_ed_vars
```
Comment (rgknox) [inline main/EDTypesMod.F90]: In the discussion at the top of the PR, you mentioned that you were moving procedure specifications to the end after the module variables were defined.  I have no strong feelings about order, but wanted to note it here.

File: main/FatesParameterDerivedMod.F90 [hunk 3]
```diff
@@ -14,7 +14,10 @@ module FatesParameterDerivedMod
   use FatesConstantsMod,     only : g_per_kg
   use FatesInterfaceMod,     only : nleafage
   
-  type param_derived_type
+  implicit none
```
[the remark this answers is not in the document]
Comment (rosiealice) [reply main/FatesParameterDerivedMod.F90]: This just made me laugh inappropriately loudly in a very large room full of people.

File: main/FatesParametersInterface.F90 [hunk 4]
```diff
@@ -53,19 +54,21 @@ module FatesParametersInterface
      type(parameter_type), private :: parameters(max_params)
 
    contains
```
Comment (rgknox) [inline main/FatesParametersInterface.F90]: @glemieux , in the following block of changes, I was surprised that previously listed procedures that were explicitly public were switched to undefined (no label).  Were these not used globally?  Also, private functions are listed down below, which are implicitly private now, right?

**Comment by rgknox** [comment]
@glemieux , I looked through the changes and they look good. Thanks!
I left a few comments in there, mostly just questions.  Pending discussion and conflict resolution and re-test I will integrate.

File: fire/SFParamsMod.F90 (hunk 1 above)
Comment (glemieux) [reply fire/SFParamsMod.F90]: That's a good question.  Chivers' book isn't explicit but the example code they give to show suggests that it's not the case that it's implicitly public.  I'll write up a quick script and test it out in a bit.

File: main/EDTypesMod.F90 (hunk 2 above)
Comment (glemieux) [reply main/EDTypesMod.F90]: Good catch.  I'll fix that.

File: main/FatesParametersInterface.F90 (hunk 4 above)
Comment (rgknox) [reply main/FatesParametersInterface.F90]: ok, just saw that those were bound inside a type, which is declared as public, (@glemieux showed me)

File: fire/SFParamsMod.F90 (hunk 1 above)
Comment (glemieux) [reply fire/SFParamsMod.F90]: @rgknox based on the [small sample code](https://github.com/glemieux/fates-scratch/blob/master/fortran-scratch/ProtectedTestMain.F90) I worked up it looks like setting something as `protected` does not make it implicitly `public`.

**Comment by glemieux** [comment]
Deconflicted the branch with master and re-ran tests.  All expected pass, b4b:

/glade/u/home/glemieux/scratch/clmed-tests/private-public-modules-mergedeconflict.fates.cheyenne.intel.C08fc948-F8efc24f

**Pull request merged**